moveit_kinematics/test/test_kinematics_plugin.cpp

```diff
@@ -0,0 +1,737 @@
+        << "Actual: " << expr2 << " [" << val2.w << ", " << val2.x << ", " << val2.y << ", " << val2.z << "]";
+    // clang-format on
+  }
+  testing::AssertionResult expectNearHelper(const char* expr1, const char* expr2, const char* abs_error_expr,
+                                            const std::vector<geometry_msgs::Pose>& val1,
+                                            const std::vector<geometry_msgs::Pose>& val2, double abs_error)
+  {
+    if (val1.size() != val2.size())
+      return testing::AssertionFailure() << "Different vector sizes"
+                                         << "\nExpected: " << expr1 << " (" << val1.size() << ")"
+                                         << "\nActual: " << expr2 << " (" << val2.size() << ")";
+
+    for (size_t i = 0; i < val1.size(); ++i)
+    {
+      ::std::stringstream ss;
+      ss << "[" << i << "].position";
+      GTEST_ASSERT_(isNear((expr1 + ss.str()).c_str(), (expr2 + ss.str()).c_str(), abs_error_expr, val1[i].position,
+                           val2[i].position, abs_error),
+                    GTEST_NONFATAL_FAILURE_);
+
+      ss.str("");
+      ss << "[" << i << "].orientation";
+      GTEST_ASSERT_(isNear((expr1 + ss.str()).c_str(), (expr2 + ss.str()).c_str(), abs_error_expr, val1[i].orientation,
+                           val2[i].orientation, abs_error),
+                    GTEST_NONFATAL_FAILURE_);
+    }
+    return testing::AssertionSuccess();
+  }
+
+  void searchIKCallback(const geometry_msgs::Pose& ik_pose, const std::vector<double>& joint_state,
+                        moveit_msgs::MoveItErrorCodes& error_code)
+  {
+    std::vector<std::string> link_names = { tip_link_ };
+    std::vector<geometry_msgs::Pose> poses;
+    if (!kinematics_solver_->getPositionFK(link_names, joint_state, poses))
+    {
+      error_code.val = error_code.PLANNING_FAILED;
+      return;
+    }
+
+    EXPECT_GT(poses[0].position.z, 0.0f);
+    if (poses[0].position.z > 0.0)
+      error_code.val = error_code.SUCCESS;
+    else
+      error_code.val = error_code.PLANNING_FAILED;
+  }
+
+public:
+  robot_model::RobotModelPtr robot_model_;
+  robot_model::JointModelGroup* jmg_;
+  kinematics::KinematicsBasePtr kinematics_solver_;
+  std::string root_link_;
+  std::string tip_link_;
+  std::string group_name_;
+  std::vector<std::string> joints_;
+  std::vector<double> seed_;
+  std::vector<double> consistency_limits_;
+  double timeout_;
+  int num_fk_tests_;
+  int num_ik_cb_tests_;
+  int num_ik_tests_;
+  int num_ik_multiple_tests_;
+  int num_nearest_ik_tests_;
+};
+
+#define EXPECT_NEAR_POSES(lhs, rhs, near)                                                                              \
+  SCOPED_TRACE("EXPECT_NEAR_POSES(" #lhs ", " #rhs ")");                                                               \
+  GTEST_ASSERT_(expectNearHelper(#lhs, #rhs, #near, lhs, rhs, near), GTEST_NONFATAL_FAILURE_);
```

Comment:
move to our testing utils files, see comment above

Reply:
Done.